aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -90,18 +93,125 @@ impl SessionId {
 
 pub struct SessionExt<'r, 'l, S> {
     inner: Session<'r, 'l, S>,
+    move_vm: &'l MoveVM,
+    remote: &'r S,
 }
 
 impl<'r, 'l, S> SessionExt<'r, 'l, S>
 where
     S: MoveResolverExt,
 {
-    pub fn new(inner: Session<'r, 'l, S>) -> Self {
-        Self { inner }
+    pub fn new(inner: Session<'r, 'l, S>, move_vm: &'l MoveVM, remote: &'r S) -> Self {
+        Self {
+            inner,
+            move_vm,
+            remote,
+        }
     }
 
     pub fn finish(self) -> VMResult<SessionOutput> {
         let (change_set, events, mut extensions) = self.inner.finish_with_extensions()?;
+
+        let common_error = PartialVMError::new(StatusCode::UNREACHABLE).finish(Location::Undefined);
+        let mut change_set_grouped = MoveChangeSet::new();
+        for (addr, account_changeset) in change_set.into_inner() {
+            let mut resource_groups: BTreeMap<StructTag, AccountChangeSet> = BTreeMap::new();
+            let mut account_changeset_grouped = AccountChangeSet::new();
+            let (modules, resources) = account_changeset.into_inner();
+
+            // * Separate the resource groups from the non-resource groups
+            // * non-resource groups are kept as is
+            // * resource groups are merged into the correct format as deltas to the source data
```

Comment:
Probably should split out the logic around resource groups into a different file?

Reply:
Yup, per your point above, there are definitely optimizations we should generalize on, test thoroughly and then apply whichever we can. Part of the constraint I faced is ineffective abstraction models and the pains of Rust... definitely areas to improve in the code base :)

Specifically, it wasn't sensible to route a lifetime'd MoveVM into the StorageAdapter

Reply:
I'm not entirely certain we benefit by moving the logic into another file. Sure we could write some unit tests, but we already have good testing around that harness. There's also a plethora of other code of similar nature in that file. Though ... admittedly I may be overly excited or missing something.